Make best attempt at fixing trivia

Bartleby2718 opened the pull request:
I think this is now ready for review. I had to delete some tests from #160, so I'd appreciate it if anyone could give me specific suggestions on where to improve. I agree that there are some inconsistencies in terms of where I've touched, but I believe I did need all of these to pass all tests.

Thanks in advance!

This fixes #713.

Bartleby2718 commented:
By the way, can anyone assign the ticket and the PR to me?

manfred-brands commented:
> By the way, can anyone assign the ticket and the PR to me?

Done.

Bartleby2718 commented:
All tests are passing now.

manfred-brands reviewed (changes requested):
Thanks @Bartleby2718 for your contribution.
Your changes and new tests are comprehensive.

There is, as expected, a lot of repeated:
```csharp
return (actualArgument, constraintArgument.WithTriviaFrom(actualArgument));
```
I wonder if you can drop all those and only move this to the `ClassicModelAssertUsageCodeFix`, where you already deal with the Trivia from the first and last argument.

Bartleby2718 commented on src/nunit.analyzers/ClassicModelAssertUsage/ContainsClassicModelAssertUsageCodeFix.cs:
```diff
@@ -31,7 +31,7 @@ public sealed class ContainsClassicModelAssertUsageCodeFix
 
             var actualArgument = argumentNamesToArguments[NUnitFrameworkConstants.NameOfActualParameter].WithNameColon(null);
-            return (actualArgument, constraintArgument);
+            return (actualArgument, constraintArgument.WithTriviaFrom(actualArgument));
```
@manfred-brands I think this is where I'd need your help here. I had tried to move it to `ClassicModelAssertUsageCodeFix`, but it appeared to me that there is no one simple solution that covers all assertions.

- If you see https://github.com/nunit/nunit.analyzers/blob/master/src/nunit.analyzers/ClassicModelAssertUsage/ClassicModelAssertUsageCodeFix.cs#L60-L68, we get a _new_ `actual` argument and a _new_ `constraint` argument from `ConstructActualAndConstraintArguments`, which is a `virtual` method of `ClassicModelAssertUsageCodeFix`.
- However, within the implementation of `ConstructActualAndConstraintArguments` in the subclasses of `ClassicModelAssertUsageCodeFix` (like this class), this `actualArgument` before the `return` is the _old_ argument.
- `ClassicModelAssertUsageCodeFix` does have the mapping `argumentNamesToArguments`, but it doesn't know the where the trivia of the new arguments should come from.
  - For most assertions, the parameter names are `expected` and `actual`.
  - For `ClassicAssert.Greater`, the parameter names are `arg1` and `arg2`.
  - For `ClassicAssert.IsTrue`, the parameter name is `condition`.
- Therefore, it appears that handling trivia in the individual codefixes would be the only way, but then it would be easy to miss/forget when adding a new one. As such, I wasn't entirely sure where the logic should live.
  - Or... maybe it's okay because there won't be any more `ClassicModelAssertUsageCodeFix`es.

**Given the above, do you think the trivia handling logic should live in `ClassicModelAssertUsageCodeFix` or its implementation?**

Bartleby2718 replied:
Edit: I think I got this.

Bartleby2718 commented on src/nunit.analyzers/ClassicModelAssertUsage/ClassicModelAssertUsageCodeFix.cs:
```diff
@@ -77,6 +77,13 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
                 newArguments.Add(interpolatedMessageArgument);
 
+            // Fix trailing trivia for the first and the last argument
+            if (newArguments.Count > 1)
+                newArguments[0] = newArguments[0].WithoutTrailingTrivia();
```
`newArgument.Count > 1` is necessary here. Consider the following assertion:
```cs
ClassicAssert.True(
    true
);
```
(Note: this is already covered by an existing test case.)

Will keep this conversation unresolved until I decide where to keep the trivia handling logic (i.e. in individual CodeFixers vs. `ClassicModelAssertUsageCodeFix`): https://github.com/nunit/nunit.analyzers/pull/754#discussion_r1632373149

Bartleby2718 replied:
Ended up centralizing the trivia handling logic to `ClassicModelAssertUsageCodeFix`.

Bartleby2718 commented on src/nunit.analyzers/ClassicModelAssertUsage/AreEqualClassicModelAssertUsageCodeFix.cs:
```diff
@@ -42,11 +42,11 @@ public sealed class AreEqualClassicModelAssertUsageCodeFix
 
             var actualArgument = argumentNamesToArguments[NUnitFrameworkConstants.NameOfActualParameter].WithNameColon(null);
-            return (actualArgument, SyntaxFactory.Argument(equalToInvocationNode));
+            return (actualArgument, SyntaxFactory.Argument(equalToInvocationNode).WithTriviaFrom(actualArgument));
```
Ah, I did think it was weird that I didn't need this to pass all tests, and this was exactly the test I was missing. I can play around this test case a little more.

Bartleby2718 replied:
This was very helpful. I've centralized the trivia handling logic to `ClassicModelAssertUsageCodeFix`, so this is no longer applicable. This part is no longer touched.

Bartleby2718 commented on src/nunit.analyzers/ConstraintUsage/SomeItemsConstraintUsageCodeFix.cs:
```diff
@@ -23,7 +23,12 @@ protected override (ExpressionSyntax? actual, ExpressionSyntax? constraintExpres
+                    actual,
+                    constraintExpression is not null
+                        ? constraintExpression.WithTriviaFrom(conditionNode)
+                        : constraintExpression);
```
I like this! However, I've centralized the trivia handling logic to `ClassicModelAssertUsageCodeFix`, so this is no longer applicable. This part is no longer touched.

Bartleby2718 commented:
@manfred-brands I've:
- added some test cases (`CodeFixMaintainsReasonableTriviaWithAllArgumentsOnSameLine`) to cover https://github.com/nunit/nunit.analyzers/pull/754#discussion_r1632182772. (For assertions like `ClassicAssert.IsEmpty` where only one argument is enough, I've added a `message` argument so that there are more than one arguments, all on the same line.)
- centralized the trivia-handling logic, allowing me to revert some unnecessary changes

I feel much better about the changes now. Thanks for the suggestions!

manfred-brands commented on src/nunit.analyzers/Extensions/ArgumentListSyntaxExtensions.cs:
```diff
@@ -36,5 +51,37 @@ internal static class ArgumentListSyntaxExtensions
+                }
+            }
+
+            foreach (var separator in separators)
```
Could add an `else`  here to prevent creating an enumerator for an array with no elements.

manfred-brands commented on src/nunit.analyzers/ConstraintUsage/BaseConditionConstraintCodeFix.cs:
```diff
@@ -126,13 +127,26 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
-            var newArgumentsList = assertNode.ArgumentList.WithArguments(newArguments);
+            var newArgumentsList = assertNode.ArgumentList
+                .WithArguments(newArguments)
+                .WithAdditionalAnnotations(Formatter.Annotation);
```
The latter should be added to the `WithArguments` extension method as that method destroys any existing Trivia if it is not a new-line and therefore requires formatting.

manfred-brands commented on src/nunit.analyzers.tests/ClassicModelAssertUsage/AreEqualClassicModelAssertUsageCodeFixTests.cs:
```diff
@@ -295,15 +295,61 @@ public void Test(object actual, object expected)
-        public void CodeFixPreservesLineBreakBeforeMessage()
+        public void CodeFixMaintainsReasonableTriviaWithEndOfLineClosingParen([Values] bool hasMessage)
+        {
+            var commaAndMessage = hasMessage
+                ? @",
+                ""message"""
+                : string.Empty;
```
Not sure if the below is clearer or not.

```suggestion
            var commaAndMessage = hasMessage ? ",\n\"message\"" : string.Empty;
```

Bartleby2718 replied:
@manfred-brands Yeah I didn't like this part of the code either.

So the spaces before the `""message""` do matter, so if we were to _not_ use a verbatim string, it should be:
```cs
            var commaAndMessage = hasMessage
                ? ",\r\n                \"message\""
                : string.Empty;
```
Do you find this more readable than the old code? Happy to make the change everywhere if you think so. (Note that the same applies to https://github.com/nunit/nunit.analyzers/pull/754#discussion_r1634236689)

manfred-brands replied:
Both are not ideal, but my preference would be the latter. It clearly indicates a newline inside a string. From the previous one I missed the importance of the spaces.

manfred-brands commented on src/nunit.analyzers.tests/ConstraintsUsage/StringConstraintUsageCodeFixTests.cs:
```diff
@@ -121,5 +130,58 @@ public void AnalyzeStringBooleanMethodAssertThatIsFalse(string method, string an
+        {
+            var analyzerId = NegativeAssertDictionary[method][1];
+            var suggestedConstraint = NegativeAssertDictionary[method][2];
+            var optionalNewline = newlineBeforeClosingParen ? @"
+            " : string.Empty;
```
Doesn't have to use a literal string here.

```suggestion
            var optionalNewline = newlineBeforeClosingParen ? "\n" : string.Empty;
```

manfred-brands reviewed (approved):
Thanks @Bartleby2718. This looks good now.

Bartleby2718 commented:
@mikkelbu Could you review this one when you get a chance?

mikkelbu commented:
Sorry for the slow answer, but I've had very little spare time the last couple of weeks. I should have time to look at it tonight (in 8-9 hours time)

mikkelbu commented on src/nunit.analyzers.tests/ClassicModelAssertUsage/AreSameClassicModelAssertUsageCodeFixTests.cs:
```diff
@@ -133,5 +133,94 @@ public void TestMethod()
+            var expected = new object();
+            var actual = new object();
+
+            ↓ClassicAssert.AreSame(
+                expected,
+                actual{optionalNewline});
+        }}");
+            var fixedCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
+        public void TestMethod()
+        {{
+            var expected = new object();
+            var actual = new object();
+
+            Assert.That(
+                actual,
+                Is.SameAs(expected){optionalNewline});
+        }}");
```
These are not on the same line
```suggestion
            ↓ClassicAssert.AreSame(
                expected, actual{optionalNewline});
        }}");
            var fixedCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
        public void TestMethod()
        {{
            var expected = new object();
            var actual = new object();

            Assert.That(
                actual, Is.SameAs(expected){optionalNewline});
        }}");
```

mikkelbu commented on src/nunit.analyzers.tests/ClassicModelAssertUsage/ContainsClassicModelAssertUsageCodeFixTests.cs:
```diff
@@ -133,5 +133,94 @@ public void TestMethod()
+            var instance = new object();
+            var collection = Array.Empty<object>();
+
+            ↓ClassicAssert.Contains(
+                instance,
+                collection{optionalNewline});
+        }}");
+            var fixedCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
+        public void TestMethod()
+        {{
+            var instance = new object();
+            var collection = Array.Empty<object>();
+
+            Assert.That(
+                collection,
+                Does.Contain(instance){optionalNewline});
```
All arguments are not on the same line
```suggestion
            ↓ClassicAssert.Contains(
                instance, collection{optionalNewline});
        }}");
            var fixedCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
        public void TestMethod()
        {{
            var instance = new object();
            var collection = Array.Empty<object>();

            Assert.That(
                collection, Does.Contain(instance){optionalNewline});
```

mikkelbu reviewed (commented):
Looks great @Bartleby2718. I've only added a handful of some minor nit-pick comments, but the overall PR is great. Thanks for the contribution 👍.

Bartleby2718 commented on src/nunit.analyzers.tests/ConstraintsUsage/StringConstraintUsageCodeFixTests.cs:
```diff
@@ -19,11 +21,18 @@ public class StringConstraintUsageCodeFixTests
+            { nameof(string.EndsWith), new[] { nameof(string.EndsWith), AnalyzerIdentifiers.StringEndsWithConstraintUsage, "Does.Not.EndWith" } },
+        };
+        private static readonly object[] NegativeAssertData = NegativeAssertDictionary.Values.ToArray();
+        private static readonly string[] StringConstraints = new[]
```
That technically works, but I didn't want `StringConstraints` to reference `Negative*`, so I've modified this a little bit.

Bartleby2718 commented:
@mikkelbu This should now be ready for another review!

mikkelbu commented:
Sorry about leaving this hanging. I'll take a look at this next weekend.

mikkelbu reviewed (approved):
Looks great. Thanks for this huge effort @Bartleby2718. I'll merge this now
